<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>LCMS-Seq by szostaklab</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>LCMS-Seq</h1>
        <p>Automated sequencing by two-dimensional analysis of mass chromatograms</p>

        <p class="view"><a href="https://github.com/szostaklab/lcmsseq">View the Project on GitHub <small>szostaklab/lcmsseq</small></a></p>


        <ul>
          <li><a href="https://github.com/szostaklab/lcmsseq/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/szostaklab/lcmsseq/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/szostaklab/lcmsseq">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <p><img src="https://raw.githubusercontent.com/szostaklab/lcmsseq/gh-pages/Github_graphic.png" alt=""></p>

<h3>
<a id="welcome-to-the-szostak-labs-lcms-seq-code-repository" class="anchor" href="#welcome-to-the-szostak-labs-lcms-seq-code-repository" aria-hidden="true"><span class="octicon octicon-link"></span></a>Welcome to the Szostak Lab's LCMS-Seq code repository</h3>

<p>This code demonstrates an algorithm to automate sequence extraction from laddering patterns in liquid chromatography-mass spectrometry data. The algorithm has been implemented as a Python 3 package, accompanying our manuscript <em>Bidirectional direct sequencing of non-canonical RNA by two-dimensional analysis of mass chromatograms</em> published in XXX 2015 in the <em>Journal of the American Chemical Society</em>. The source code has been made available under the GNU AGPL v3.</p>

<h3>
<a id="documentation-and-installation" class="anchor" href="#documentation-and-installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation and installation</h3>

<p>Please see the publication for background on the algorithm and application.  The <a href="https://github.com/szostaklab/lcmsseq/blob/master/README.md">readme</a> file, found in the repo top directory, contains detailed instructions for installation on multiple platforms.  The package has been lightly tested on Windows, Linux, and OS X using Python 3.4.  There are several dependencies, including the clustalo binary that must be installed prior to use.  Please see the readme for details.  Several example datasets are also included in the /examples directory, and instructions on running the examples are also included in the readme file.</p>

<h3>
<a id="authors" class="anchor" href="#authors" aria-hidden="true"><span class="octicon octicon-link"></span></a>Authors</h3>

<p>Anders Bj√∂rkbom, Victor S. Lelyveld, Shenglong Zhang, Weicheng Zhang, Chun Pong Tam, J. Craig Blain, and Jack W. Szostak at HHMI, MGH, and Harvard</p>

<h3>
<a id="help" class="anchor" href="#help" aria-hidden="true"><span class="octicon octicon-link"></span></a>Help</h3>

<p>Please contact <a href="https://github.com/lelyveld" class="user-mention">@lelyveld</a> for help or pull requests or open an issue.</p>

<hr>

<h2>
<a id="getting-started-with-the-lcmsseq-python-package" class="anchor" href="#getting-started-with-the-lcmsseq-python-package" aria-hidden="true"><span class="octicon octicon-link"></span></a>Getting started with the lcmsseq Python package</h2>

<p>Note: An updated version of this documentation is available in the <a href="https://github.com/szostaklab/lcmsseq/blob/master/README.md">readme</a>.</p>

<h2>
<a id="short-version" class="anchor" href="#short-version" aria-hidden="true"><span class="octicon octicon-link"></span></a>Short version</h2>

<p>Change to the directory where you've cloned the repository.  In Debian/Ubuntu Linux, do:</p>

<pre><code>$ sudo pip3 install numpy matplotlib pandas scipy patsy statsmodels
...
$ sudo apt-get install clustalo
...
$ sudo python3 setup.py install
...
$ cd examples
...
$ python3 test.py
...
</code></pre>

<p>The final command will process and display an example dataset and resulting sequencing reads.</p>

<h2>
<a id="long-version" class="anchor" href="#long-version" aria-hidden="true"><span class="octicon octicon-link"></span></a>Long version</h2>

<p>This package gives a proof-of-concept implementation of an algorithm for extracting sequence information from a
two-dimensional dataset of degradative fragments of a sequenceable polymer.</p>

<h3>
<a id="requirements" class="anchor" href="#requirements" aria-hidden="true"><span class="octicon octicon-link"></span></a>Requirements</h3>

<p>You must have Python 3. In recent Linux distributions, Python 3 is typically installed by default. If not, use your
 distro's package manager to install the latest python 3 package.</p>

<p>A few generally available Python 3 packages (most from the Scipy stack) must also be installed:</p>

<pre><code>numpy 1.9+
matplotlib 1.3+
statsmodels 0.6+
scipy 0.9+
pandas 0.7.1+
patsy 0.3+
</code></pre>

<p>The bottom three packages are only necessary to satisfy installation requirements for statsmodels. Information
about the Scipy stack may be found here:</p>

<pre><code>http://www.scipy.org/install.html
</code></pre>

<p>On Linux, it should generally be possible to use pip to satisfy all Python 3 module dependencies quickly. Or use your
favorite python package manager. Alternatively, it may be possible to install certain Python distributions that contain
all or part of the above packages (e.g. Anaconda). Installation details for each environment are suggested below.</p>

<p>For this initial release, the package currently makes use of the clustalo binary for alignment operations.  See specific
instructions below. The binary for ClustalO must be named 'clustalo' must be executable from the current path for the
current user.</p>

<p>Finally, for this release, the input data must be generated by Agilent MassHunter Qualitative Analysis software.
See below under <em>Input Data</em> for more discussion.</p>

<h4>
<a id="satisfying-requirements-in-linux" class="anchor" href="#satisfying-requirements-in-linux" aria-hidden="true"><span class="octicon octicon-link"></span></a>SATISFYING REQUIREMENTS IN LINUX:</h4>

<p>On Debian/Ubuntu systems with typical configuration, it should be possible to run the following command to install all
required python packages:</p>

<pre><code>$ sudo pip3 install numpy matplotlib scipy pandas patsy statsmodels
</code></pre>

<p>On Ubuntu-based systems that use the apt package manager, ClustalO can easily be obtained by running the following
command to install it from default Ubuntu repositories.</p>

<pre><code>$ sudo apt-get install clustalo
</code></pre>

<p>If you obtain clustalo from another source, the binary for ClustalO must be named 'clustalo' must be executable from the
 current path for the current user.</p>

<p>After fulfilling these requirements, see the installation section below to install this package.</p>

<h4>
<a id="satisfying-requirements-in-os-x" class="anchor" href="#satisfying-requirements-in-os-x" aria-hidden="true"><span class="octicon octicon-link"></span></a>SATISFYING REQUIREMENTS IN OS X:</h4>

<p>Users who already have the Scipy stack for Python 3 installed may be able to simply run the following:</p>

<pre><code>sudo pip3 install patsy statsmodels
</code></pre>

<p>Alternatively, if this is your first installation of Python 3 and/or the Scipy stack, the easiest method to satisfy
package dependencies in OS X is to install a free scientific Python 3 distribution, such as Anaconda. This suite can be
downloaded from here:</p>

<pre><code>http://continuum.io/downloads#py34
</code></pre>

<p>If this works for you, then skip to installation of clustalo below.  Alternatively, download and install the latest
stable Python 3.4+ binary release obtainable from <a href="https://www.python.org">https://www.python.org</a> . The current version at the time of this
writing is here:</p>

<pre><code>https://www.python.org/ftp/python/3.4.3/python-3.4.3-macosx10.6.pkg
</code></pre>

<p>Installing the packages may be achieved through pip3, as above for Linux. However, as of this writing, direct
compilation of numpy is problematic.</p>

<p>The OS X binary for clustalo can be found here:</p>

<pre><code>http://www.clustal.org/omega/clustal-omega-1.2.0-macosx
</code></pre>

<p>The binary should be renamed, made executable, and moved to an accessible path, as follows:</p>

<pre><code>$ mv clustal-omega-1.2.0-macosx clustalo
$ chmod a+x clustalo
$ sudo mv clustalo /usr/local/bin
</code></pre>

<p>After fulfilling these requirements, see the installation section below to install this package.</p>

<h4>
<a id="satisfying-requirements-in-windows" class="anchor" href="#satisfying-requirements-in-windows" aria-hidden="true"><span class="octicon octicon-link"></span></a>SATISFYING REQUIREMENTS IN WINDOWS</h4>

<p>For Windows users, the struggle is real. It may be possible satisfy all dependencies by installing a third-party Python
distribution. At present, the easiest strategy in Windows is to install Anaconda's Python 3 distribution, available
here:</p>

<pre><code>http://continuum.io/downloads#py34
</code></pre>

<p>If Python 3 from this distribution works for you, then skip down to clustalo installation. If not, download and
install the latest stable Python 3.4+ binary release obtainable from <a href="https://www.python.org">https://www.python.org</a>. During installation in
Windows, select the option to "add python.exe to path," which will make life much easier. The
current Windows 64-bit installer can be found here:</p>

<pre><code>https://www.python.org/ftp/python/3.4.3/python-3.4.3.amd64.msi
</code></pre>

<p>To install packages, generally pip does not work without errors. Thankfully, Christoph Gohlke at the Laboratory for
Fluorescence Dynamics, University of California, Irvine, maintains an unofficial repository of Windows python package
binaries, such that direct compilation on Windows is unnecessary. After installing Python 3.4 in Windows, download the
cp34 versions of the required packages. Be sure to install the bit version that matches your version of Python 3.4
(simply run python at the command line to check whether you have the 32-bit or 64-bit version). The pip "wheels" below
allow quick installation in Windows. Currently, those files can be found at the following URL:</p>

<pre><code>http://www.lfd.uci.edu/~gohlke/pythonlibs/
</code></pre>

<p>For 64-bit Python 3.4 in 64-bit Windows, download the following files:</p>

<pre><code>numpy‚Äë1.9.2+mkl‚Äëcp34‚Äënone‚Äëwin_amd64.whl
scipy‚Äë0.15.1‚Äëcp34‚Äënone‚Äëwin_amd64.whl
matplotlib‚Äë1.4.3‚Äëcp34‚Äënone‚Äëwin_amd64.whl
pandas‚Äë0.16.0‚Äëcp34‚Äënone‚Äëwin_amd64.whl
patsy‚Äë0.3.0‚Äëpy2.py3‚Äënone‚Äëany.whl
statsmodels‚Äë0.6.1‚Äëcp34‚Äënone‚Äëwin_amd64.whl
</code></pre>

<p>Install each of these packages by running the following command on each filename from a command prompt:</p>

<pre><code>pip install filename.whl
</code></pre>

<p>The ClustalO binary must be installed and accessible in the current path. ClustalO for Windows can be obtained here:</p>

<pre><code>http://www.clustal.org/omega/clustal-omega-1.2.0-win32.zip
</code></pre>

<p>Unzip the contents of this zipfile to a good location, such as C:\clustalo . The path must be changed to allow
clustalo.exe to be accessible from elsewhere. The path can be changed from Control Panel &gt; System &gt; Advanced system
settings &gt; Advanced tab &gt; Environment Variables &gt; System Variables &gt; Path. Append "C:\clustalo;" to the value of the
"Path" variable and hit OK.</p>

<p>After fulfilling these requirements, see the installation section below to install this package.</p>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h3>

<p>A setup script is included with this package for installation.  From the root directory of the package, simply invoke
the setup.py script. In Linux or OS X, run:</p>

<pre><code>$ sudo python3 setup.py install
</code></pre>

<p>In Windows, run:</p>

<pre><code>python setup.py install
</code></pre>

<h3>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h3>

<p>Several instructive examples are included in the examples subdirectory. Run these from the command line by typing:</p>

<pre><code>$ cd examples
$ python3 test.py
$ python3 figures.py
</code></pre>

<p>From wtihin the Python 3 interpreter, the following commands are all that is necessary to run the example dataset
provided in the examples/ subdirectory from the python interpreter.</p>

<pre><code>&gt;&gt;&gt; import lcmsseq
&gt;&gt;&gt; lcmsseq.read_params('default.cfg')
&gt;&gt;&gt; lcmsseq.process('compounds.csv')
</code></pre>

<p>For the time being, processing parameters are currently accessible from a parameter CSV file. The default parameters
 can be found in the bundled file "default.cfg". An explanation of the function of each parameter is given as a comment
 in this configuration file.</p>

<p>Alternatively, the modules can be run from the command line as follows:</p>

<pre><code>$ python3 lcmsseq.py dataset.csv parameters.cfg
</code></pre>

<p>Parameters:</p>

<pre><code>&lt;dataset.csv&gt;: compound dataset filename
Filename of a comma-delimited list of compounds extracted from LCMS data by, e.g., Agilent MassHunter Qualitative
 Analysis Find by Molecular Feature. The format of this file is described below under _Input Data_.

&lt;parameters.cfg&gt;: parameters filename
Filename of a comma-delimited list of parameters key,value pairs. The bundled filed default.cfg shows the
expected content of this file.
</code></pre>

<p>The parameter file specifies the location of two key comma-delimited databases, containing a list of adducts and bases.
The adduct database is a csv file containing a list of adduct exact masses for clustering. The base database is a csv
file containing chemical formulas that correspond to mass differences between degradative fragments differing by loss
of a single RNA monomer. Further instructions are available in files examples/adducts.csv and examples/bases.csv . An
Excel file is also included to aid in generating the base database starting from nucleoside chemical formulas.</p>

<h3>
<a id="input-data" class="anchor" href="#input-data" aria-hidden="true"><span class="octicon octicon-link"></span></a>Input Data</h3>

<p>At the time of this initial release, the package has been designed to import comma-delimited (csv) files generated by
Agilent MassHunter Qualitative Analysis software. Specifically, the file must be a list of compounds identified
using the "Find by Molecular Feature" algorithm. Guidelines on using this feature to generate useful datasets can
be found in our initial publication. Other compound-finding software are likely to be suitable, but this package
currently expects to find several Agilent-generated column headings in the imported csv file.</p>

<p>Use the CSV export file from Agilent MassHunter Qualitative Analysis software, or ensure your CSV file is a list of
compounds with one compound per row and contains (at the absolute minimum) the following column headers:</p>

<pre><code>Cpd,Mass,RT,Vol
</code></pre>

<p>These columns contain the following information:</p>

<pre><code>Cpd: compound ID (generally an integer)
Mass: neutral mass
RT: retention time
Vol: integrated intensity
</code></pre>

<p>Processing will fail if any of these columns are missing. The package can also filter on a number of other optional
column headers that are typically assigned by Agilent's software, although this filtering is not strictly required.</p>

<h3>
<a id="additional-comments" class="anchor" href="#additional-comments" aria-hidden="true"><span class="octicon octicon-link"></span></a>Additional comments</h3>

<p>The module makes use of the python 3 multiprocessing package to spawn several parallel processes to generate walk
trajectories across the dataset. For this release, multicore features are operational in Linux and OS X environments.</p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/szostaklab">szostaklab</a></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-69070519-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
